crates/ruff_python_formatter/src/expression/string/docstring.rs

```diff
@@ -1156,6 +1188,214 @@ impl<'src> CodeExampleRst<'src> {
+#[derive(Debug)]
+struct CodeExampleMarkdown<'src> {
+    /// The lines that have been seen so far that make up the block.
+    lines: Vec<CodeExampleLine<'src>>,
+
+    /// The indent of the line "opening" fence of this block measured via
+    /// `indentation_length`.
+    ///
+    /// This indentation is trimmed from the indentation of every line in the
+    /// body of the code block,
+    opening_fence_indent: TextSize,
+
+    /// The kind of fence, backticks or tildes, used for this block. We need to
+    /// keep track of which kind was used to open the block in order to look
+    /// for a correct close of the block.
+    fence_kind: MarkdownFenceKind,
+
+    /// The size of the fence, in codepoints, in the opening line. A correct
+    /// close of the fence must use *at least* this many characters.
+    fence_len: usize,
+}
+
+impl<'src> CodeExampleMarkdown<'src> {
+    /// Looks for the start of a Markdown [fenced code block].
+    ///
+    /// If the start of a block is found, then this returns a correctly
+    /// initialized Markdown code block. Callers should print the line as given
+    /// as it is not retained as part of the block.
+    ///
+    /// [fenced code block]: https://spec.commonmark.org/0.30/#fenced-code-blocks
+    fn new(original: InputDocstringLine<'src>) -> Option<CodeExampleMarkdown<'src>> {
+        let (opening_fence_indent, rest) = indent_with_suffix(original.line);
+        // Quit quickly in the vast majority of cases.
+        if !rest.starts_with("```") && !rest.starts_with("~~~") {
+            return None;
+        }
+
+        static FENCE_START: Lazy<Regex> = Lazy::new(|| {
+            Regex::new(
+                r"(?xm)
+                ^
+                (?:
+                    # In the backtick case, info strings (following the fence)
+                    # cannot contain backticks themselves, since it would
+                    # introduce ambiguity with parsing inline code. In other
+                    # words, if we didn't specifically exclude matching `
+                    # in the info string for backtick fences, then we might
+                    # erroneously consider something to be a code fence block
+                    # that is actually inline code.
+                    #
+                    # NOTE: The `ticklang` and `tildlang` capture groups are
+                    # currently unused, but there was some discussion about not
+                    # assuming unlabeled blocks were Python. At the time of
+                    # writing, we do assume unlabeled blocks are Python, but
+                    # one could inspect the `ticklang` and `tildlang` capture
+                    # groups to determine whether the block is labeled or not.
+                    (?<ticks>```+)(?:\s*(?<ticklang>(?i:python|py|python3|py3))[^`]*)?
+                    |
+                    (?<tilds>~~~+)(?:\s*(?<tildlang>(?i:python|py|python3|py3))\p{any}*)?
+                )
```

Comment:
Is there a source for this regex or was it written by hand for these purposes?

Reply:
Is there any sense in checking for the triple quotes / tildes out-of-band before running the regex, since the vast majority of lines won't contain them? Or would you expect (wink wink) regex to perform equivalently-or-better?

Edit: hah, nevermind, you already have this above 🤦